pkg/nas/nodeserver_test.go

```diff
@@ -0,0 +1,202 @@
+		})
+	}
+}
+
+func TestFallbackCNFSAndRecord(t *testing.T) {
+	client := fake.NewSimpleClientset(&v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "pod1",
+			Namespace: "default",
+		},
+	})
+	req := &csi.NodePublishVolumeRequest{
+		VolumeContext: map[string]string{
+			utils.PodNameKey:      "pod1",
+			utils.PodNamespaceKey: "default",
+		},
+	}
+	eventRecorder := record.NewFakeRecorder(5)
+	ctx, _ := utils.WithPodInfo(context.Background(), client, req)
+
+	primaryCNFSName, fallbackCNFSName := "primary", "fallback"
+	tests := []struct {
+		name           string
+		primaryCNFS    *v1beta1.ContainerNetworkFileSystem
+		fallbackCNFS   *v1beta1.ContainerNetworkFileSystem
+		expectFallback bool
+		expectErr      bool
+	}{
+		{
+			name:           "Always fallback strategy",
+			primaryCNFS:    fakeCNFS(primaryCNFSName, "", "", fallbackCNFSName, v1beta1.FallbackStrategyAlways),
+			fallbackCNFS:   fakeCNFS(fallbackCNFSName, "", "", "", ""),
+			expectFallback: true,
+		},
+		{
+			name:           "IfConnectFailed fallback strategy",
+			primaryCNFS:    fakeCNFS(primaryCNFSName, "", "", fallbackCNFSName, v1beta1.FallbackStrategyIfConnectFailed),
+			fallbackCNFS:   fakeCNFS(fallbackCNFSName, "", "", "", ""),
+			expectFallback: true,
+		},
+		{
+			name:           "IfMountTargetUnhealthy fallback strategy",
+			primaryCNFS:    fakeCNFS(primaryCNFSName, v1beta1.StatusUnavailable, "", fallbackCNFSName, v1beta1.FallbackStrategyIfMountTargetUnhealthy),
+			fallbackCNFS:   fakeCNFS(fallbackCNFSName, "", "", "", ""),
+			expectFallback: true,
+		},
+		{
+			name:         "Non-existent fallback CNFS",
+			primaryCNFS:  fakeCNFS("primary", "", "", "non-existent-cnfs", ""),
+			fallbackCNFS: fakeCNFS(fallbackCNFSName, "", "", "", ""),
+			expectErr:    true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			cnfsGetter := newFakeCNFSGetter(tt.fallbackCNFS)
+			server := nodeServer{
+				config: &internal.NodeConfig{
+					KubeClient: client,
+					CNFSGetter: cnfsGetter,
+				},
+				recorder: eventRecorder,
+			}
+			actual, err := server.fallbackCNFSAndRecord(ctx, req, tt.primaryCNFS)
+			if tt.expectErr {
+				assert.Error(t, err)
+			} else {
+				if tt.expectFallback {
+					assert.Equal(t, *tt.fallbackCNFS, *actual)
+					assert.Len(t, eventRecorder.Events, 1)
+
+					oldCNFSName, newCNFSName := tt.primaryCNFS.Name, tt.fallbackCNFS.Name
+					msg := <-eventRecorder.Events
+					switch tt.primaryCNFS.Spec.Fallback.Strategy {
+					case v1beta1.FallbackStrategyAlways:
+						assert.Contains(t, msg, fmt.Sprintf(cnfsAlwaysFallbackEventTmpl, oldCNFSName, newCNFSName))
+					case v1beta1.FallbackStrategyIfConnectFailed:
+						assert.Contains(t, msg, fmt.Sprintf(cnfsIfConnectFailedFallbackEventTmpl, oldCNFSName, newCNFSName))
+					case v1beta1.FallbackStrategyIfMountTargetUnhealthy:
+						assert.Contains(t, msg, fmt.Sprintf(cnfsIfMountTargetUnhealthyFallbackEventTmpl, oldCNFSName, newCNFSName))
+					}
```

Comment:
How about add a expectedEvent field to `tests`, instead of copy pasting a large segment of code?